Add fi_info utility to libfabric.

pmmccorm opened the pull request:
It works same as it did in fabtests:

```
  -n, --node=NAME               node name or address
  -p, --port=PNUM               port number
  -c, --caps=CAP1|CAP2..        one or more capabilities: FI_MSG|FI_RMA...
  -m, --mode=MOD1|MOD2..        one or more modes, default all modes
  -e, --ep_type=EPTYPE          specify single endpoint type: FI_EP_MSG, FI_EP_DGRAM...
  -a, --addr_format=FMT         specify accepted address format: FI_FORMAT_UNSPEC, FI_SOCKADDR...
  -f, --provider=PROV           specify provider explicitly
  -v, --verbose                 verbose output
  --version                     print version info and exit
```

But now by default prints abbreviated info by default:

```
verbs: IB-0x80fe
    version: 1.0
    type: FI_EP_MSG
    protocol: FI_PROTO_RDMA_CM_IB_RC
sockets: IP
    version: 1.0
    type: FI_EP_MSG
    protocol: FI_PROTO_SOCK_TCP
sockets: IP
    version: 1.0
    type: FI_EP_DGRAM
    protocol: FI_PROTO_SOCK_TCP
sockets: IP
    version: 1.0
    type: FI_EP_RDM
    protocol: FI_PROTO_SOCK_TCP
```

Use -v or --verbose to see the old verbose output.

Signed-off-by: pmmccorm <patrick.m.mccormick@intel.com>

jsquyres commented:
Probably worth a separate PR to add the output of the env variables (that will require a new accessor in src/var.c).

bturrubiates commented on configure.ac:
```diff
@@ -70,6 +70,7 @@ AC_ARG_ENABLE([direct],
 
 dnl Checks for programs
 AC_PROG_CC_C99
+AM_PROG_CC_C_O
```
Which version of automake are we recommending? `AM_PROG_CC_C_O` is obsolescent in newer versions and automake will print a warning about the configury containing the macro.

pmmccorm replied:
Whichever one works..which one should it be here?

bturrubiates replied:
So `AM_PROG_CC_C_O` is obsolescent in Automake 1.14 and above because the `AC_PROG_CC` macro handles the same checks. In Automake 1.14 and above `AM_PROG_CC_C_O` is just a no-op. It isn't scheduled for removal until Automake 2.0. I don't know what the appropriate solution is.

goodell replied:
It's worth caring about, but I think the fix is just to remove our usage of `AM_PROG_CC_C_O` from `configure.in`.  Ben, can you try making that change and building a fresh copy of libfabric on one of our RHEL6 nodes that has automake-1.11.1?  Note that you'll probably have to `module unload cisco/autotools` and check with `automake --version` to make sure you are really testing with the stock autotools from the RHEL distro.  Assuming we don't get any warnings from automake about omitting that macro, then that should be the easy fix.

If automake complains about missing `AM_PROG_CC_C_O` for versions <1.14 but complains about its presence for versions >=1.14 then we'll have to do something trickier or require 1.14 as the minimum (undesirable, IMO, but not the end of the world).

bturrubiates replied:
@pmmccorm I've tested with various am versions and I don't think it's an issue right now. We can't easily remove it, but the versions of automake > 1.14 that we have are not generating warnings/errors. We'll handle this at some later time when it's more pressing.

pmmccorm replied:
Sounds good, thanks.

pmmccorm commented:
Updated. Added var printing system, output looks like:

```
FI_PSM_AM_MSG=1 ./src/util/fi_info -e

# sockets: Drop every Nth dgram frame (debug only)
# FI_SOCKETS_DGRAM_DROP_RATE

# sockets: How many miliseconds to spin while waiting for progress
# FI_SOCKETS_PE_WAITTIME

# psm: Whether to check PSM version number compatibility
# FI_PSM_CHECK_VERSION

# psm: Unique Job ID required by the PSM fabric
# FI_PSM_UUID

# psm: Whether to use tagged messages for large size RMA or not (default: yes)
# FI_PSM_TAGGED_RMA

# psm: Whether to use active message based messaging or not (default: no)
FI_PSM_AM_MSG=1

# psm: Whether to turn on the name server or not
# FI_PSM_NAME_SERVER
```

Had to fixup fi_var_* API to get things working though. I put those changes in separate commit, same PR.

pmmccorm commented:
A couple of notes:

fi_getinfo() has to be called before we can list the env vars. We can change this so that a call to _fi_getvars() triggers the initialization (similar to how getinfo and fi_fabric do). So if all the providers registered their vars in their INI functions (vs closer to their use in their getinfo functions) we could list the env vars without potentially initializing the hardware. If this is desired speak now.

A separate issue should probably be opened to specify how the config file will look. I think it would be nice if it looked very similar to the printout that "fi_info -e" gives. Do we need more information in it than that (a list of var=value)?

shefty commented:
My preference is for the providers to init their variables as early as possible, but read them immediately before use.  And to read them each time they are used (within reason), so that changes get picked up.

The output looks good.  If anything I would change the help text to indicate what values are accepted.  e.g. replace "whether to do something" with "set to yes or 1 to enable something"

shefty commented on include/rdma/fi_var.h:
```diff
@@ -40,6 +40,16 @@
+	char *env_var_name;
+
+	struct fi_var_t *next;
+};
```
Unless these structures are allocated per call to fi_getvars(), this structure is either racy or limits what can be done internally.  There's no way for libfabric to modify this list safely if an app can be walking through these structures because of a lack of synchronization.

We need a different approach for iterating over this data.  Regardless of how we intend this data to be used, it is exposed through the public API.  An app can (and eventually will) walk this list and look at these structures.

jsquyres commented on src/var.c:
```diff
@@ -41,20 +41,18 @@
 static struct fi_var_t *fi_vars = NULL;
 
+__attribute__((visibility ("default")))
+struct fi_var_t *DEFAULT_SYMVER_PRE(_fi_getvars)(void) {
```
Aren't symbols with an underscore prefix in the namespace of the compiler / OS?

pmmccorm replied:
Yes probably. This is a public facing API but it should only be used by tooling or otherwise "special" users. Not sure if there is a way we can communicate that, or maybe we don't.

shefty replied:
> Yes probably. This is a public facing API but it should only be used by
> tooling or otherwise "special" users.

We need to design this with the expectation that some random app is going to try to use it for some obscure reason.  E.g. looking for some specific option that they can enable.

We definitely don't want to provide handles to internal data structures.  Maybe we need something more like:

struct fi_vars {
	const char *prov_name;
	const char *var_name;
	...
}

int fi_getvars(struct fi_vars *vars, int *count);
fi_freevars(vars);

We could drop fi_freevars() if we want the caller to allocate the memory, with the necessary size returned by fi_getvars().

jsquyres commented:
@pmmccorm Good suggestion about registering in the INI functions.  ...mmm, are there any cases where the INI functions are invoked pre-main?

shefty commented:
I'm creating new patches that simply moves fi_info from fabtests to libfabric, with no other changes to fi_info.  Separate patches should adjust the output and then add support for displaying the environment variables.

pmmccorm commented:
@shefty I have this done and working, I can break this up.

shefty commented:
Ok - I tried to pull in the patch that just moves fi_info, but it fails to build because it depends on the var changes.  I was just going to move fi_info into the source tree, with no other changes (except what's needed to build).  From your patch, I would move info.c from src/util to just util.

pmmccorm commented:
Rebased and refactored to just add fi_info to the tree. Installs under /usr/bin and is part of the libfabric RPM
